### PR TITLE
Attend user from wait list after someone from main list cancel visit request

### PR DESCRIPTION
**Pull request by avozniak**
Resolves [github issue #240 ](https://github.com/pivorakmeetup/pivorak-web-app/issues/240)

### Description

Set first user from waiting list to main list when user from main list cancel visit request

### How to test instructions

Create event.

Set `limit total` and `limit verified` to 1

Event should have `registration` status

---

Login as verified user

Should see `Attend` link at event page

Click `Attend` link

Should see successful confirm message `You have been confirmed for this event!`

Should see `Cancel attendance` link

Logout

---

Login as another verified user

Should see `Seems like all seats have been 
taken, but you can add yourself to waiting list`

Should see `Attend` link

After click on `Attend link` shoud see `Your event request is under review. we'll get back to you asap` message.

Shoud see `Cancel attedance` link

Click at `Cancel attedance` link

Should see `Sorry, but have canceled your visit request`

### Pre-merge checklist
 
- [x] The PR relates to a single subject with a clear title and description
- [x] Verify that feature branch is up-to-date with `development` (if not - rebase it). 

### Screenshots

| Before                                      | After                                       |
| ------------------------------------------- | ------------------------------------------- |
| alt text| alt|

### CODE REVIEW

**Review by sveredyuk — COMMENTED** [review]
Please, agree this feature with @pivorak

File: app/controllers/admin/visit_requests_controller.rb [hunk 1]
```diff
@@ -15,7 +15,7 @@ def event
     end
 
     def visit_requests
-      @visit_requests ||= search_against(base_scope).includes(:user, :event).order(:id)
```
Comment (denys-medynskyi) [inline app/controllers/admin/visit_requests_controller.rb:18]: Why do you need to change this line?

Comment (avozniak) [reply app/controllers/admin/visit_requests_controller.rb:18]: I remove only `:event` parameter from includes. Whith them there is a n+1  request

File: app/controllers/visit_requests_controller.rb [hunk 2]
```diff
@@ -14,7 +14,7 @@ def create
   end
 
   def destroy
-    visit_request.destroy
+    VisitRequest::Cancel.call(visit_request)
```
Comment (denys-medynskyi) [inline app/controllers/visit_requests_controller.rb:17]: Do not make changes that are not related to current issue

Comment (avozniak) [reply app/controllers/visit_requests_controller.rb:17]: oh, sorry, I forgot, I should have make separated PR for this

File: app/services/visit_request/cancel.rb [hunk 3]
```diff
@@ -6,10 +6,23 @@ def initialize(visit_request)
...
       visit_request.canceled!
+      attend_first_user_from_waitlist
     end
 
     private
 
     attr_reader :visit_request
+
+    def attend_first_user_from_waitlist
+      if waiting_list_request = VisitRequest.where(waiting_list: true).first
```
Comment (denys-medynskyi) [inline app/services/visit_request/cancel.rb]: 1. First `if` can be replaced with [guard clause](https://devblast.com/b/what-are-guard-clauses)
2. `VisitRequest.where(waiting_list: true).first` -> `VisitRequest. waiting_list.first`
3. `VisitRequest. waiting_list.first` -> we should supply the ordering, because PostgreSQL can return records randomly. Here is an [example from stackoverflow](https://stackoverflow.com/questions/9060875/activerecordqueryingfirst-returns-third-member-of-the-collection).

File: app/services/visit_request/cancel.rb [hunk 4]
```diff
@@ -6,10 +6,23 @@ def initialize(visit_request)
...
+        waiting_list_request.waiting_list = false
+        waiting_list_request.save
+        if visit_request.event.status == 'registration'
```
Comment (denys-medynskyi) [inline app/services/visit_request/cancel.rb]: You can do `event.registration?` because of [this](https://github.com/pivorakmeetup/pivorak-web-app/blob/development/app/models/event.rb#L16)

File: app/services/visit_request/cancel.rb [hunk 5]
```diff
@@ -6,10 +6,23 @@ def initialize(visit_request)
...
+      attend_first_user_from_waitlist
     end
 
     private
 
     attr_reader :visit_request
+
+    def attend_first_user_from_waitlist
+      if waiting_list_request = VisitRequest.where(waiting_list: true).first
+        waiting_list_request.waiting_list = false
```
Comment (denys-medynskyi) [inline app/services/visit_request/cancel.rb]: You can replace 
```
waiting_list_request.waiting_list = false
waiting_list_request.save
``` 
with one line:
`waiting_list_request.main_list!` [defined here](https://github.com/pivorakmeetup/pivorak-web-app/blob/development/app/models/visit_request.rb#L24)

**Review by denys-medynskyi — CHANGES_REQUESTED** [review]
Fix & respond to my suggestions

File: app/services/visit_request/cancel.rb [hunk 6]
```diff
@@ -6,10 +6,23 @@ def initialize(visit_request)
...
     end
 
     private
 
     attr_reader :visit_request
+
+    def attend_first_user_from_waitlist
+      return unless waiting_list_request = VisitRequest.waiting_list.first
+
+      waiting_list_request.main_list!
```
Comment (denys-medynskyi) [inline app/services/visit_request/cancel.rb:19]: Consider such situation:
```
VisitRequest::Approve
```
raises error, but it will not rollback `waiting_list_request` to the waiting list.
These 2 operations should be in transactions.

Please add such spec:
1. stub VisitRequest::Approve to raise an error
2. request waiting list should be rollbacked to the `false` state

**Review by denys-medynskyi — CHANGES_REQUESTED** [review]
Checkout what I found

File: app/controllers/admin/visit_requests_controller.rb (hunk 1 above)
Comment (denys-medynskyi) [reply app/controllers/admin/visit_requests_controller.rb:18]: Can you add tit as a separate PR too?
As a part of - https://github.com/pivorakmeetup/pivorak-web-app/issues/379

File: app/services/visit_request/cancel.rb [hunk 7]
```diff
@@ -6,10 +6,23 @@ def initialize(visit_request)
...
       visit_request.canceled!
+      attend_first_user_from_waitlist
     end
 
     private
 
     attr_reader :visit_request
+
+    def attend_first_user_from_waitlist
+      return unless waiting_list_request = VisitRequest.waiting_list.first
```
Comment (denys-medynskyi) [inline app/services/visit_request/cancel.rb:17]: Please add `.order(created_at: :asc).first`.
As I said already PostgreSQL can return records randomly.

File: app/services/visit_request/cancel.rb [hunk 8]
```diff
@@ -6,10 +6,23 @@ def initialize(visit_request)
...
+
+      waiting_list_request.main_list!
+
+      if visit_request.event.registration?
+        VisitRequest::Approve.call(waiting_list_request)
+      else
+        VisitRequest::FinalConfirmation(waiting_list_request)
+      end
+    end
```
Comment (denys-medynskyi) [inline app/services/visit_request/cancel.rb:26]: As you see here -> https://codecov.io/gh/pivorakmeetup/pivorak-web-app/pull/417/src/app/services/visit_request/cancel.rb?before=app/services/visit_request/cancel.rb
This branch is not covered by specs.
You can add feature spec for an event with the different status.

**Comment by denys-medynskyi** [comment]
Regarding the branch/repo.
I mean this branch should be made from github.com/pivorakmeetup/pivorak-web-app not from your forked repo.
screen shot 2017-07-11 at 10 48 23 pm

**Comment by denys-medynskyi** [comment]
Also PR title/branch should be as in [contribution section](https://github.com/pivorakmeetup/pivorak-web-app#contribution)
